rectools/models/sasrec.py

```diff
@@ -922,57 +920,103 @@ def on_train_start(self) -> None:
+    def training_step(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
+        """TODO"""
         if self.loss == "softmax":
-            # We are using CrossEntropyLoss with a multi-dimensional case
-
-            # Logits must be passed in form of [batch_size, n_items + n_special_tokens, session_max_len],
-            #  where n_items + n_special_tokens is number of classes
-
-            # Target label indexes must be passed in a form of [batch_size, session_max_len]
-            # (`0` index for "PAD" ix excluded from loss)
-
-            # Loss output will have a shape of [batch_size, session_max_len]
-            # and will have zeros for every `0` target label
-
-            loss = torch.nn.functional.cross_entropy(
-                logits.transpose(1, 2), y, ignore_index=0, reduction="none"
-            )  # [batch_size, session_max_len]
-            loss = loss * w
-            n = (loss > 0).to(loss.dtype)
-            loss = torch.sum(loss) / torch.sum(n)
-            return loss
-        raise ValueError(f"loss {loss} is not supported")
+            return self._calc_softmax_loss(batch)
+        if self.loss == "BCE":
+            return self._calc_bce_loss(batch)
+        if self.loss == "gBCE":
+            return self._calc_gbce_loss(batch)
+
+        raise ValueError(f"loss {self.loss} is not supported")
+
+    def _calc_softmax_loss(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
+        x, y, w = batch["x"], batch["y"], batch["yw"]
+        item_embs, session_embs = self.forward(x)
+        # We are using CrossEntropyLoss with a multi-dimensional case
+
+        # Logits must be passed in form of [batch_size, n_items + n_item_extra_tokens, session_max_len],
+        #  where n_items + n_item_extra_tokens is number of classes
+
+        # Target label indexes must be passed in a form of [batch_size, session_max_len]
+        # (`0` index for "PAD" ix excluded from loss)
+
+        # Loss output will have a shape of [batch_size, session_max_len]
+        # and will have zeros for every `0` target label
+        logits = session_embs @ item_embs.T
+        loss = torch.nn.functional.cross_entropy(
+            logits.transpose(1, 2), y, ignore_index=0, reduction="none"
+        )  # [batch_size, session_max_len]
+        loss = loss * w
+        n = (loss > 0).to(loss.dtype)
+        loss = torch.sum(loss) / torch.sum(n)
+        return loss
+
+    def _calc_bce_loss(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
+        x, y, w, negatives = batch["x"], batch["y"], batch["yw"], batch["negatives"]
+        # [n_items + n_item_extra_tokens, n_factors], [batch_size, session_max_len, n_factors]
+        item_embs, session_embs = self.forward(x)
+        pos_neg = torch.cat([y.unsqueeze(-1), negatives], dim=-1)  # [batch_size, session_max_len, n_negatives + 1]
+        item_embs = item_embs[pos_neg]  # [batch_size, session_max_len, n_negatives + n_item_extra_tokens, n_factors]
```

Comment:
I guess it's `[batch_size, session_max_len, n_negatives + 1, n_factors]`

Reply:
fixed